src/v/cluster/tests/internal_secret_test.cc

```diff
@@ -0,0 +1,145 @@
+// Copyright 2020 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "cluster/feature_manager.h"
+#include "cluster/internal_secret.h"
+#include "cluster/internal_secret_frontend.h"
+#include "cluster/tests/cluster_test_fixture.h"
+#include "cluster/types.h"
+#include "features/feature_table.h"
+#include "model/metadata.h"
+#include "test_utils/async.h"
+#include "test_utils/fixture.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/sleep.hh>
+#include <seastar/util/defer.hh>
+#include <seastar/util/log.hh>
+
+#include <algorithm>
+#include <iterator>
+#include <system_error>
+#include <vector>
+
+using namespace std::chrono_literals;
+using namespace cluster;
+
+namespace {
+
+// Generate successive size_t starting from init
+constexpr auto iota = [](size_t init = 0) {
+    return [i = init]() mutable { return i++; };
+};
+
+constexpr auto fetch_secret_from =
+  [](application* app, internal_secret::key_t const& key) {
+      return app->controller->get_internal_secret_frontend()
+        .local()
+        .fetch(key, 5s)
+        .get();
+  };
+
+}; // namespace
+
+FIXTURE_TEST(test_internal_secret_fetch, cluster_test_fixture) {
+    std::vector<application*> apps{
+      create_node_application(model::node_id{0}),
+      create_node_application(model::node_id{1}),
+      create_node_application(model::node_id{2})};
+
+    wait_for_all_members(3s).get();
+
+    tests::cooperative_spin_wait_with_timeout(10s, [&apps] {
+        return apps.front()->controller->get_feature_table().local().is_active(
+          features::feature::internal_secrets);
+    }).get();
+
+    // Create a key via each application
+    std::vector<internal_secret::key_t> keys;
+    keys.reserve(apps.size());
+    std::generate_n(
+      std::back_inserter(keys), apps.size(), [index{iota()}]() mutable {
+          const auto i = index();
+          return internal_secret::key_t{fmt::format("n{}", i).c_str()};
+      });
+
+    std::vector<fetch_internal_secret_reply> replies;
+    replies.reserve(apps.size());
+
+    // Fetch (create) a secret from each application
+    std::generate_n(
+      std::back_inserter(replies), apps.size(), [&, index{iota()}]() mutable {
+          const auto i = index();
+          return fetch_secret_from(apps[i], keys[i]);
+      });
+
+    // validate the secrets were created successfuly and that the keys match
+    for (size_t i{0}; i < apps.size(); ++i) {
+        BOOST_REQUIRE_EQUAL(replies[i].ec, errc::success);
+        BOOST_REQUIRE_EQUAL(replies[i].secret.key()(), keys[i]);
+        BOOST_REQUIRE_EQUAL(replies[i].secret.value()().length(), 128);
+    }
+
+    // Check secrets are all different
```

Comment:
Bit confused: I would have thought the fetch() call always going to get routed through to the leader, so we should get the same secret each time?  And then further down on line 98 we do another round of fetches where we assert that the results of fetch()s are all the same, whereas here we asserted each fetch() gave a different result.

Reply:
The secrets get cached locally on each shard. Rob found a bug where all the secrets are actually empty (and the same!) on the second request. 

There's a bit to this PR that's probably missing, and that's the cache invalidation, which would go through the request cycle again.

Inside `internal_secret_frontend::fetch`
```
        if (reply.ec == errc::success) {
            co_await _store.invoke_on_all([secret{reply.secret}](auto& store) {
                store.insert_or_assign(secret);
            });
        }
```